pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,470 @@
+	}
+
+	m.isIPv4Enabled, _ = config.IsIPv4Enabled(nodeConfig, trafficEncapMode)
+	m.isIPv6Enabled, _ = config.IsIPv6Enabled(nodeConfig, trafficEncapMode)
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    m.onNodeAdd,
+		UpdateFunc: m.onNodeUpdate,
+		DeleteFunc: m.onNodeDelete,
+	})
+
+	nlmInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    m.onNodeLatencyMonitorAdd,
+		UpdateFunc: m.onNodeLatencyMonitorUpdate,
+		DeleteFunc: m.onNodeLatencyMonitorDelete,
+	})
+
+	return m
+}
+
+// Is current node
+func (m *NodeLatencyMonitor) isCurrentNode(node *corev1.Node) bool {
+	return node.Name == m.nodeName
+}
+
+// onNodeAdd is the event handler for adding Node.
+func (m *NodeLatencyMonitor) onNodeAdd(obj interface{}) {
+	node := obj.(*corev1.Node)
+	if m.isCurrentNode(node) {
+		return
+	}
+
+	m.latencyStore.addNode(node)
+
+	klog.V(4).InfoS("Node added", "Node", klog.KObj(node))
+}
+
+// onNodeUpdate is the event handler for updating Node.
+func (m *NodeLatencyMonitor) onNodeUpdate(oldObj, newObj interface{}) {
+	node := newObj.(*corev1.Node)
+	if m.isCurrentNode(node) {
+		return
+	}
+
+	m.latencyStore.updateNode(node)
+
+	klog.V(4).InfoS("Node updated", "Node", klog.KObj(node))
+}
+
+// onNodeDelete is the event handler for deleting Node.
+func (m *NodeLatencyMonitor) onNodeDelete(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+			return
+		}
+		node, ok = deletedState.Obj.(*corev1.Node)
+		if m.isCurrentNode(node) {
+			return
+		}
+
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Node object", "obj", deletedState.Obj)
+			return
+		}
+	}
+
+	m.latencyStore.deleteNode(node)
+}
+
+// onNodeLatencyMonitorAdd is the event handler for adding NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorAdd(obj interface{}) {
+	nlm := obj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor added", "NodeLatencyMonitor", klog.KObj(nlm))
+
+	m.updateLatencyConfig(nlm)
+}
+
+// onNodeLatencyMonitorUpdate is the event handler for updating NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorUpdate(oldObj, newObj interface{}) {
+	oldNLM := oldObj.(*v1alpha1.NodeLatencyMonitor)
+	newNLM := newObj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	m.updateLatencyConfig(newNLM)
+}
+
+// updateLatencyConfig updates the latency config based on the NodeLatencyMonitor CRD.
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) {
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
```

Comment:
There is a data race between `updateLatencyConfig` and `monitorLoop`, the former of which updates the variable while the latter reads the variable, without lock acquired.

Reply:
Thanks, there is indeed a problem with race here, I'll add a lock here.